File: lib/geoscript/feature/field.js
var UTIL = require("../util");
var PROJ = require("../proj");
var GEOM = require("../geom");
var register = require("./util").register;
var Factory = require("../factory").Factory;

var jts = Packages.com.vividsolutions.jts;
var geotools = Packages.org.geotools;
var AttributeTypeBuilder = geotools.feature.AttributeTypeBuilder;
var types = {};

// map type names to java types
var javaTypeNames = ["String", "Integer", "Short", "Float", "Long", "Double"];
javaTypeNames.forEach(function(str) {
    var type = java.lang[str];
    types[str] = type;
    types[type] = type;
});

// map type names to jts geometry types
var jtsTypeNames = ["Geometry", "Point", "LineString", "Polygon", "MultiPoint", "MultiLineString", "MultiPolygon"];
jtsTypeNames.forEach(function(str) {
    var type = jts.geom[str];
    types[str] = type;
    types[type] = type;
});

var getTypeName = function(type) {
    var name;
    for (var str in types) {
        if (types[str] === type) {
            name = str;
            break;
        };
    }
    if (!name) {
        throw new Error("Can't get name for field type: " + type);
    }
    return name;
};

var getType = function(value) {
    var name;
    // deal with geometries
    if (value instanceof GEOM.Geometry) {
        name = value._geometry.getGeometryType();
    } else {
        var t = typeof value;
        if (t === "string") {
            name = "String";
        } else if (t === "number") {
            name = "Double";
        }
    }
    var type = types[name];
    if (!type) {
        if (name) {
            throw new Error("Can't resolve field type name: " + name);
        } else {
            throw new Error("Unsupported field type: " + value);
        }
    }
    return type;
};

var Field = UTIL.extend(Object, {

    constructor: function Field(config) {
        if (config) {
            if (!config.name) {
                throw new Error("Field config must include 'name' property.");
            }
            if (!config.type) {
                throw new Error("Field config must include 'type' property.");
            }
            if (!types[config.type]) {
                throw new Error("Unsupported field type: " + config.type);
            }
            var builder = new AttributeTypeBuilder();
            builder.setName(config.type);
            builder.setBinding(types[config.type]);
            var projection = config.projection;
            if (projection) {
                if (!(projection instanceof PROJ.Projection)) {
                    projection = new PROJ.Projection(projection);
                }
                builder.setCRS(projection._projection);
            }
            builder.setMinOccurs(config.minOccurs || 0);
            builder.setMaxOccurs(config.maxOccurs || 1);
            if ("isNillable" in config) {
                builder.setNillable(!!config.isNillable);
            }
            if ("defaultValue" in config) {
                // TODO: pass java values
                builder.setDefaultValue(config.defaultValue);
            }
            this._field = builder.buildDescriptor(config.name);
        }
    },
    
    get name() {
        return String(this._field.getLocalName());
    },
    
    get type() {
        return getTypeName(this._field.getType().getBinding());
    },
    
    get minOccurs() {
        return Number(this._field.getMinOccurs());
    },
    
    get maxOccurs() {
        return Number(this._field.getMaxOccurs());
    },
    
    get isNillable() {
        return Boolean(this._field.getIsNillable());
    },
    
    get defaultValue() {
        return this._field.getDefaultValue();
    },
    
    get projection() {
        var projection;
        if (GEOM[this.type]) {
            var _projection = this._field.getCoordinateReferenceSystem();
            if (_projection) {
                projection = PROJ.Projection.from_(_projection);
            }
        }
        return projection;
    },
    
    equals: function(field) {
        return Boolean(this._field.equals(field._field));
    },
    
    /** private: method[toFullString]
     */
    toFullString: function() {
        return "name: \"" + this.name + "\", type: " + this.type;
    }

});

Field.fromValue = function(name, value) {
    var config = {
        name: name, 
        type: getTypeName(getType(value))
    };
    if (value instanceof GEOM.Geometry && value.projection) {
        config.projection = value.projection;
    }
    return new Field(config);
};

Field.from_ = function(_field) {
    var field = new Field();
    field._field = _field;
    return field;
};

exports.Field = Field;

// register a field factory for the module
register(new Factory(Field));

File: src/main/java/org/geoscript/js/feature/Field.java
package org.geoscript.js.feature;

import org.geoscript.js.GeoObject;
import org.geoscript.js.proj.Projection;
import org.geotools.feature.AttributeTypeBuilder;
import org.mozilla.javascript.Context;
import org.mozilla.javascript.Function;
import org.mozilla.javascript.NativeObject;
import org.mozilla.javascript.ScriptRuntime;
import org.mozilla.javascript.Scriptable;
import org.mozilla.javascript.Wrapper;
import org.mozilla.javascript.annotations.JSConstructor;
import org.mozilla.javascript.annotations.JSGetter;
import org.mozilla.javascript.annotations.JSStaticFunction;
import org.opengis.feature.type.AttributeDescriptor;
import org.opengis.feature.type.GeometryDescriptor;
import org.opengis.referencing.crs.CoordinateReferenceSystem;

public class Field extends GeoObject implements Wrapper {

    /** serialVersionUID */
    private static final long serialVersionUID = -45254255795575119L;
    
    /**
     * Optional title.
     */
    private String title;

    private AttributeDescriptor descriptor;
    
    /**
     * Prototype constructor.
     */
    public Field() {
    }
    /**
     * Constructor from AttributeDescriptor.
     * @param scope
     * @param crs
     */
    public Field(Scriptable scope, AttributeDescriptor descriptor) {
        this.setParentScope(scope);
        this.setPrototype(Module.getClassPrototype(Field.class));
        this.descriptor = descriptor;
    }

    /**
     * Constructor from NativeObject (from Java).
     * @param scope
     * @param config
     */
    public Field(Scriptable scope, NativeObject config) {
        this(config);
        this.setParentScope(scope);
        this.setPrototype(Module.getClassPrototype(Field.class));
    }

    private Field(NativeObject config) {
        String name = (String) getRequiredMember(config, "name", String.class);
        String typeName = (String) getRequiredMember(config, "type", String.class);
        Type type;
        try {
            type = Type.valueOf(typeName);
        } catch (IllegalArgumentException e) {
            throw ScriptRuntime.constructError("Error", "Unsupported field type: " + typeName);
        }
        AttributeTypeBuilder builder = new AttributeTypeBuilder();
        builder.setName(name);
        builder.setBinding(type.getBinding());

        Object descObj = getOptionalMember(config, "description", String.class);
        if (descObj != null) {
            builder.setDescription((String) descObj);
        }

        CoordinateReferenceSystem crs = null;
        if (config.has("projection", config)) {
            Object projObj = config.get("projection", config);
            if (projObj instanceof Projection) {
                crs = ((Projection) projObj).unwrap();
            } else if (projObj instanceof String) {
                crs = new Projection(getParentScope(), (String) projObj).unwrap();
            } else if (projObj != null){
                throw ScriptRuntime.constructError("Error", "Invalid projection object.");
            }
        }
        if (crs != null) {
            builder.setCRS(crs);
        }

        Object minOccursObj = getOptionalMember(config, "minOccurs", Number.class);
        int minOccurs = -1;
        if (minOccursObj != null) {
            minOccurs = ((Number) minOccursObj).intValue();
            if (minOccurs < -1) {
                throw ScriptRuntime.constructError("Error", "Invalid minOccurs value: " + Context.toString(minOccursObj));
            }
            builder.setMinOccurs(minOccurs);
        }

        Object maxOccursObj = getOptionalMember(config, "maxOccurs", Number.class);
        int maxOccurs = -1;
        if (maxOccursObj instanceof Number) {
            maxOccurs = ((Number) maxOccursObj).intValue();
            if (maxOccurs < minOccurs || maxOccurs == 0) {
                throw ScriptRuntime.constructError("Error", "Invalid maxOccurs value: " + Context.toString(maxOccursObj));
            }
            builder.setMaxOccurs(maxOccurs);
        } else if (minOccurs > -1) {
            maxOccurs = minOccurs == 0 ? 1 : minOccurs;
            builder.setMaxOccurs(maxOccurs);
        }

        Object isNillableObj = getOptionalMember(config, "isNillable", Boolean.class);
        if (isNillableObj != null) {
            builder.setNillable((Boolean) isNillableObj);
        }

        if (config.has("defaultValue", config)) {
            Object defaultValue = config.get("defaultValue", config);
            if (defaultValue != null) {
                builder.setDefaultValue(jsToJava(defaultValue));
            }
        }

        descriptor = builder.buildDescriptor(name);
    }
    
    @JSConstructor
    public static Object constructor(Context cx, Object[] args, Function ctorObj, boolean inNewExpr) {
        if (args.length != 1) {
            throw ScriptRuntime.constructError("Error", "Constructor takes a single argument");
        }
        Field field = null;
        Object arg = args[0];
        if (arg instanceof NativeObject) {
            NativeObject config = (NativeObject) arg;
            if (inNewExpr) {
                field = new Field(config);
            } else {
                field = new Field(config.getParentScope(), config);
            }
        } else {
            throw ScriptRuntime.constructError("Error", "Could not create field from argument: " + Context.toString(arg));
        }
        return field;
    }
    
    @JSGetter
    public String getTitle() {
        return title;
    }
    
    @JSGetter
    public String getName() {
        return descriptor.getLocalName();
    }
    
    @JSGetter
    public String getDescription()  {
        return descriptor.getType().getDescription().toString();
    }
    
    @JSGetter
    public String getType() {
        Class<?> binding = descriptor.getType().getBinding();
        return Type.getName(binding);
    }
    
    @JSGetter
    public int getMinOccurs() {
        return descriptor.getMinOccurs();
    }

    @JSGetter
    public int getMaxOccurs() {
        return descriptor.getMaxOccurs();
    }

    @JSGetter
    public boolean getIsNillable() {
        return descriptor.isNillable();
    }
    
    @JSGetter
    public Object getDefaultValue() {
        return descriptor.getDefaultValue();
    }
    
    @JSGetter
    public Projection getProjection() {
        Projection projection = null;
        if (descriptor instanceof GeometryDescriptor) {
            CoordinateReferenceSystem crs = ((GeometryDescriptor) descriptor).getCoordinateReferenceSystem();
            if (crs != null) {
                projection = new Projection(getParentScope(), crs);
            }
        }
        return projection;
    }

    @JSGetter
    public Scriptable getConfig() {
        Scriptable config = super.getConfig();
        Context cx = getCurrentContext();
        Scriptable def = cx.newObject(getParentScope());
        def.put("name", def, getName());
        def.put("type", def, getType());
        Projection projection = getProjection();
        if (projection != null) {
            def.put("projection", def, projection.getId());
        }
        config.put("def", config, def);
        return config;
    }

    public Object unwrap() {
        return descriptor;
    }
    
    /**
     * Determine the string type for a given value.  Defined as a static method 
     * on the JavaScript constructor.
     */
    @JSStaticFunction
    public static String getTypeName(Object value) {
        value = jsToJava(value);
        String typeName = null;
        if (value != null) {
            typeName = Type.getName(value.getClass());
        }
        return typeName;
    }
    
    public String toFullString() {
        return "name: \"" + getName() + "\" type: " + getType();
    }


}
